bin/varnishd/cache/cache_http.c

```diff
@@ -821,36 +821,158 @@ http_GetHdrField(const struct http *hp, hdr_t hdr,
 
 /*--------------------------------------------------------------------*/
 
+static ssize_t
+http_parse_uint(const char *b, const char **e)
+{
+	ssize_t u;
+	unsigned n;
+
+	u = 0;
+	if (!vct_isdigit(*b))
+		return (-1);
+	for (; vct_isdigit(*b); b++) {
+		if (u > (SSIZE_MAX / 10))
+			return (-1);
+		u *= 10;
+		n = *b - '0';
+		if (u > (SSIZE_MAX - n))
+			return (-2);
+		u += n;
+	}
+
+	*e = b;
+	return (u);
+}
+
 ssize_t
 http_GetContentLength(const struct http *hp)
 {
 	ssize_t cl;
 	const char *b;
-	unsigned n;
 
 	CHECK_OBJ_NOTNULL(hp, HTTP_MAGIC);
 
 	if (!http_GetHdr(hp, H_Content_Length, &b))
 		return (-1);
-	cl = 0;
-	if (!vct_isdigit(*b))
+	cl = http_parse_uint(b, &b);
+	if (cl < 0)
 		return (-2);
-	for (; vct_isdigit(*b); b++) {
-		if (cl > (SSIZE_MAX / 10))
+	while (vct_islws(*b))
+		b++;
+	if (*b != '\0')
+		return (-2);
+	return (cl);
+}
+
+ssize_t
+http_GetContentRange(const struct http *hp, ssize_t *lo, ssize_t *hi)
+{
+	ssize_t tmp, cl;
+	const char *b;
+
+	CHECK_OBJ_NOTNULL(hp, HTTP_MAGIC);
+
+	if (lo == NULL)
+		lo = &tmp;
+	if (hi == NULL)
+		hi = &tmp;
+
+	*lo = *hi = -1;
+
+	if (!http_GetHdr(hp, H_Content_Range, &b))
+		return (-1);
+
+	if (strncasecmp("bytes", b, strlen("bytes")))
+		return (-1);		// Unknown range unit, ignore
+	b += strlen("bytes");
+
+	if (!vct_islws(*b))
+		return (-1);		// Unknown range unit, ignore
+	while (vct_islws(*b))
+		b++;
+	if (*b == '*') {		// Content-Range: bytes */123
+		*lo = *hi = -1;
+		b++;
+	} else {			// Content-Range: bytes 1-2/3
+		*lo = http_parse_uint(b, &b);
+		if (*lo < 0)
 			return (-2);
-		cl *= 10;
-		n = *b - '0';
-		if (cl > (SSIZE_MAX - n))
+		if (*b != '-')
+			return (-2);
+		*hi = http_parse_uint(b + 1, &b);
+		if (*hi < 0)
+			return (-2);
+	}
+	if (*b != '/')
+		return (-2);
+	if (b[1] == '*') {		// Content-Range: bytes 1-2/*
+		cl = -1;
+		b += 2;
+	} else {
+		cl = http_parse_uint(b + 1, &b);
+		if (cl <= 0)
 			return (-2);
-		cl += n;
 	}
 	while (vct_islws(*b))
 		b++;
 	if (*b != '\0')
 		return (-2);
+	if (*lo > *hi)
+		return (-2);
+	assert(cl >= -1);
+	if (cl == -1)
+		return (-1);
+	if (*lo >= cl || *hi >= cl)
+		return (-2);
+	AN(cl);
 	return (cl);
 }
 
+const char *
+http_GetRange(const struct http *hp, ssize_t *lo, ssize_t *hi)
+{
+	ssize_t tmp;
+	const char *b;
+
+	CHECK_OBJ_NOTNULL(hp, HTTP_MAGIC);
+
+	if (lo == NULL)
+		lo = &tmp;
+	if (hi == NULL)
+		hi = &tmp;
+
+	*lo = *hi = -1;
+
+	if (!http_GetHdr(hp, H_Range, &b))
+		return (NULL);
```

Comment:
hm. Should we really return "success" for no range header?

Reply:
Think of this as its own function, if there is no range, then low and high arguments will be negative. It may be used by code other than `vrg_range_init()`, and `vrg_range_init()` has its own assertion that it is only ever called when there is a range header.